SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp

```diff
@@ -0,0 +1,979 @@
+	{
+		UE_LOG(LogTemp, Warning, TEXT("Unsupported dynamic add component op received - component ID: %u entity ID: %lld"),
+			Op.data.component_id, Op.entity_id);
+		return;
+	}
+
+	TSharedPtr<improbable::Component> Data;
+
+	switch (Op.data.component_id)
+	{
+	case ENTITY_ACL_COMPONENT_ID:
+		Data = MakeShared<improbable::EntityAcl>(Op.data);
+		break;
+	case METADATA_COMPONENT_ID:
+		Data = MakeShared<improbable::Metadata>(Op.data);
+		break;
+	case POSITION_COMPONENT_ID:
+		Data = MakeShared<improbable::Position>(Op.data);
+		break;
+	case PERSISTENCE_COMPONENT_ID:
+		Data = MakeShared<improbable::Persistence>(Op.data);
+		break;
+	case ROTATION_COMPONENT_ID:
+		Data = MakeShared<improbable::Rotation>(Op.data);
+		break;
+	case UNREAL_METADATA_COMPONENT_ID:
+		Data = MakeShared<improbable::UnrealMetadata>(Op.data);
+		break;
+	case SpatialConstants::GLOBAL_STATE_MANAGER_COMPONENT_ID:
+		GlobalStateManager->ApplyData(Op.data);
+		GlobalStateManager->LinkExistingSingletonActors();
+		return;
+	default:
+		Data = MakeShared<improbable::DynamicComponent>(Op.data);
+		break;
+	}
+
+	PendingAddComponents.Emplace(Op.entity_id, Op.data.component_id, Data);
+}
+
+void USpatialReceiver::OnRemoveEntity(Worker_RemoveEntityOp& Op)
+{
+	UE_LOG(LogTemp, Log, TEXT("RemoveEntity: %lld"), Op.entity_id);
+
+	PackageMap->RemoveEntitySubobjects(Op.entity_id);
+	RemoveActor(Op.entity_id);
+}
+
+void USpatialReceiver::OnAuthorityChange(Worker_AuthorityChangeOp& Op)
+{
+	if (Op.component_id == SpatialConstants::GLOBAL_STATE_MANAGER_COMPONENT_ID
+		&& Op.authority == WORKER_AUTHORITY_AUTHORITATIVE)
+	{
+		GlobalStateManager->ExecuteInitialSingletonActorReplication();
+	}
+}
+
+void USpatialReceiver::CreateActor(Worker_EntityId EntityId)
+{
+	checkf(World, TEXT("We should have a world whilst processing ops."));
+	check(NetDriver);
+
+	UEntityRegistry* EntityRegistry = NetDriver->GetEntityRegistry();
+	check(EntityRegistry);
+
+	improbable::Position* Position = GetComponentData<improbable::Position>(*this, EntityId);
+	improbable::Metadata* Metadata = GetComponentData<improbable::Metadata>(*this, EntityId);
+	improbable::Rotation* Rotation = GetComponentData<improbable::Rotation>(*this, EntityId);
+
+	check(Position && Metadata);
+
+	AActor* EntityActor = EntityRegistry->GetActorFromEntityId(EntityId);
+	UE_LOG(LogTemp, Log, TEXT("!!! Checked out entity with entity ID %lld"), EntityId);
+
+	if (EntityActor)
+	{
+		UClass* ActorClass = GetNativeEntityClass(Metadata);
+
+		// Option 1
+		UE_LOG(LogTemp, Log, TEXT("Entity for core actor %s has been checked out on the worker which spawned it."), *EntityActor->GetName());
+
+		improbable::UnrealMetadata* UnrealMetadata = GetComponentData<improbable::UnrealMetadata>(*this, EntityId);
+		check(UnrealMetadata);
+
+		USpatialPackageMapClient* SpatialPackageMap = Cast<USpatialPackageMapClient>(NetDriver->GetSpatialOSNetConnection()->PackageMap);
+		check(SpatialPackageMap);
+
+		SubobjectToOffsetMap SubobjectNameToOffset;
```

Comment:
Lets create a helper function to create a `SubobjectToOffsetMap` from `UnrealMetadata`